### PR TITLE
Count right, SQLite.

### LINKED ISSUE #2432: SQLite: Use AUTOINCREMENT when possible
[AUTOINCREMENT](http://sqlite.org/autoinc.html) prevents the reuse of ROWIDs from previously deleted rows.

Without this, you could hit some weird concurrency issues. Consider the following.

```C#
using (var context1 = new Context())
using (var context2 = new Context());
{
    var alfred1 = context1.People.First(p => p.Id = 2);
    var alfred2 = context2.People.First(p => p.Id = 2);

    context1.People.Remove(alfred);
    context1.People.Add(new Person { Name = "Batman" });
    context1.SaveChanges(); // Adds Batman with Id = 2

    alfred2.LastName = "Pennyworth";
    context2.SaveChanges(); // Updates Batman!
}
```

### PR DESCRIPTION
**Pull request by natemcmaster**
Fix #2432. Uses AUTOINCREMENT keyword in SQLite to handle quirks in their rowid algorithm.

### CODE REVIEW

File: src/EntityFramework.Sqlite/Migrations/SqliteMigrationSqlGenerator.cs [hunk 1]
```diff
@@ -87,6 +87,25 @@ public virtual void Generate([NotNull] MoveDataOperation operation, [CanBeNull]
...
+        public override void Generate(CreateTableOperation operation, IModel model, SqlBatchBuilder builder)
+        {
+            // Lifts a primary key definition into the typename.
+            // This handles the quirks of creating integer primary keys using autoincrement, not default rowid behavior.
+            if (operation.PrimaryKey?.Columns.Length == 1)
+            {
+                var column = operation.Columns?.FirstOrDefault(c => c.Name == operation.PrimaryKey.Columns[0]);
+                if (column != null
+                    && (column.Type.Equals("INTEGER", StringComparison.InvariantCultureIgnoreCase)
+                        || column.Type.Equals("INT", StringComparison.InvariantCultureIgnoreCase)))
```
Comment (bricelam) [inline src/EntityFramework.Sqlite/Migrations/SqliteMigrationSqlGenerator.cs]: Must be exactly "INTEGER" any other value won't work.

Comment (bricelam) [inline src/EntityFramework.Sqlite/Migrations/SqliteMigrationSqlGenerator.cs]: We could probably just override the user-specified type with "INTEGER" if the CLR type is an integral type. I'll look into this with #2421.

File: src/EntityFramework.Sqlite/Migrations/SqliteMigrationSqlGenerator.cs [hunk 2]
```diff
@@ -87,6 +87,25 @@ public virtual void Generate([NotNull] MoveDataOperation operation, [CanBeNull]
...
+                {
+                    column.Type = "INTEGER PRIMARY KEY AUTOINCREMENT";
```
Comment (bricelam) [inline src/EntityFramework.Sqlite/Migrations/SqliteMigrationSqlGenerator.cs]: Hmm, I think we should use annotations for autoincrement and inline primary key.

File: test/EntityFramework.Sqlite.Tests/Migrations/SqliteMigrationSqlGeneratorTest.cs [hunk 3]
```diff
@@ -47,33 +63,6 @@ public void RestartSequenceOperation_not_supported()
...
-                    Type = "int",
-                    IsNullable = true,
-                    DefaultExpression = "10"
-                });
-
-            Assert.Equal(
-                @"ALTER TABLE ""People"" ADD ""Age"" int DEFAULT (10);" + EOL,
-                Sql);
-        }
-
```
Comment (bricelam) [inline test/EntityFramework.Sqlite.Tests/Migrations/SqliteMigrationSqlGeneratorTest.cs]: Why were these removed?

Comment (natemcmaster) [inline test/EntityFramework.Sqlite.Tests/Migrations/SqliteMigrationSqlGeneratorTest.cs]: They were testing for a scenario we removed several commits ago  (throwing on CURRENT_TIMESTAMP in the SQL generator). The base class still tests add column.

File: src/EntityFramework.Sqlite/Migrations/SqliteMigrationSqlGenerator.cs (hunk 1 above)
Comment (natemcmaster) [inline src/EntityFramework.Sqlite/Migrations/SqliteMigrationSqlGenerator.cs]: Right. "INT" will not work with autoincrement. The typemapper should be returning INTEGER, but this catches some cases in our testing where we specify "int" (and possibly where a user specifies "int").

File: src/EntityFramework.Sqlite/Migrations/SqliteMigrationSqlGenerator.cs (hunk 2 above)
Comment (natemcmaster) [inline src/EntityFramework.Sqlite/Migrations/SqliteMigrationSqlGenerator.cs]: To clarify, you're suggesting we apply this logic during model building instead and then use annotations to generate a different CreateTable operation? Or some other place...

File: test/EntityFramework.Sqlite.Tests/Migrations/SqliteMigrationSqlGeneratorTest.cs (hunk 3 above)
Comment (bricelam) [inline test/EntityFramework.Sqlite.Tests/Migrations/SqliteMigrationSqlGeneratorTest.cs]: We decided to pass the SQL through to the database and let it throw. These test that we're generating the "correct" (or at least valid-to-parse) SQL.

**Comment by natemcmaster** [comment]
Will update when #2431 is merged.

File: src/EntityFramework.Sqlite/Migrations/SqliteMigrationSqlGenerator.cs (hunk 1 above)
Comment (bricelam) [inline src/EntityFramework.Sqlite/Migrations/SqliteMigrationSqlGenerator.cs]: I think we should either...
1. Always override the type to be INTEGER
2. Only add AUTOINCREMENT when the type is exactly INTEGER

For now, let's detect if the type merely *contains* "INT" and override it with INTEGER. This is similar logic to how SQLite determines column affinity.

After #2421 is in, we can be even smarter an base it on the CLR type instead of the column type name.

File: src/EntityFramework.Sqlite/Migrations/SqliteMigrationSqlGenerator.cs (hunk 2 above)
Comment (bricelam) [inline src/EntityFramework.Sqlite/Migrations/SqliteMigrationSqlGenerator.cs]: To capture what we discussed yesterday:

1. Add an annotation like `Sqlite:Autoincrement = true` to the property during model diffing
2. Add an annotation like `Sqlite:InlinePrimaryKey = true` to the property as an implementation detail of the SqliteMigrationSqlGenerator

**Comment by natemcmaster** [comment]
@bricelam added the new implementation.

File: src/EntityFramework.Sqlite/Migrations/SqliteMigrationAnnotationProvider.cs [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+                    columnType = null;
+                }
+                if (columnType != null
+                    && columnType.Equals("integer", StringComparison.CurrentCultureIgnoreCase)
+                    && clrType != typeof(bool)
+                    && clrType != typeof(char)
+                    && clrType != typeof(byte)
+                    )
+                {
+                    yield return new Annotation(SqliteAnnotationNames.Prefix + SqliteAnnotationNames.Autoincrement, true);
```
Comment (bricelam) [inline src/EntityFramework.Sqlite/Migrations/SqliteMigrationAnnotationProvider.cs]: This may make more sense on the property

File: src/EntityFramework.Sqlite/Migrations/SqliteMigrationAnnotationProvider.cs [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+            }
+            base.For(key);
+        }
+
+        public override IEnumerable<IAnnotation> For(IProperty property)
+        {
+            if (property.IsPrimaryKey()
+                && property.EntityType.GetPrimaryKey().Properties.Count == 1)
+            {
+                yield return new Annotation(SqliteAnnotationNames.Prefix + SqliteAnnotationNames.InlinePrimaryKey, true);
```
Comment (bricelam) [inline src/EntityFramework.Sqlite/Migrations/SqliteMigrationAnnotationProvider.cs]: Not needed here. This can just be an implementation detail of the `SqliteMigrationSqlGenerator`.

File: src/EntityFramework.Sqlite/Migrations/SqliteMigrationAnnotationProvider.cs [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+                    && columnType.Equals("integer", StringComparison.CurrentCultureIgnoreCase)
+                    && clrType != typeof(bool)
+                    && clrType != typeof(char)
+                    && clrType != typeof(byte)
+                    )
+                {
+                    yield return new Annotation(SqliteAnnotationNames.Prefix + SqliteAnnotationNames.Autoincrement, true);
+                }
+            }
+            base.For(key);
```
Comment (bricelam) [inline src/EntityFramework.Sqlite/Migrations/SqliteMigrationAnnotationProvider.cs]: Don't call base on these.

File: src/EntityFramework.Sqlite/Migrations/SqliteMigrationSqlGenerator.cs [hunk 7]
```diff
@@ -87,6 +88,38 @@ public virtual void Generate([NotNull] MoveDataOperation operation, [CanBeNull]
...
+            var inlinePrimaryKeyAnnotation = operation.Columns
+                .FirstOrDefault(c => c.FindAnnotation(inlinePrimaryKey) != null)?[inlinePrimaryKey];
+
+            if (inlinePrimaryKeyAnnotation != null
+                && (bool)inlinePrimaryKeyAnnotation)
+            {
+                var column = operation.Columns?.FirstOrDefault(c => c.Name == operation.PrimaryKey.Columns[0]);
+                if (column != null)
+                {
+                    column.Type += " PRIMARY KEY";
```
Comment (bricelam) [inline src/EntityFramework.Sqlite/Migrations/SqliteMigrationSqlGenerator.cs]: Don't abuse the `Type` property like this. Override how the SQL gets generated in `ColumnDefinition`

File: test/EntityFramework.Sqlite.Tests/Migrations/SqliteMigrationSqlGeneratorTest.cs (hunk 3 above)
Comment (bricelam) [inline test/EntityFramework.Sqlite.Tests/Migrations/SqliteMigrationSqlGeneratorTest.cs]: Add these tests back.

File: src/EntityFramework.Sqlite/Migrations/SqliteMigrationAnnotationProvider.cs (hunk 4 above)
Comment (natemcmaster) [inline src/EntityFramework.Sqlite/Migrations/SqliteMigrationAnnotationProvider.cs]: Agreed. Moved it to property.

**Comment by natemcmaster** [comment]
@bricelam cleaned up the implementation to respond to your feedback. Better?

File: src/EntityFramework.Sqlite/Migrations/SqliteMigrationAnnotationProvider.cs [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+                }
+                catch (NotSupportedException)
+                {
+                    columnType = null;
+                }
+                if (columnType != null
+                    && columnType.Equals("integer", StringComparison.CurrentCultureIgnoreCase)
+                    && clrType != typeof(bool)
+                    && clrType != typeof(char)
+                    && clrType != typeof(byte)
```
Comment (bricelam) [inline src/EntityFramework.Sqlite/Migrations/SqliteMigrationAnnotationProvider.cs]: Does the CLR type actually matter? If it's an INTEGER column, things should "just work" for any CLR type.

File: src/EntityFramework.Sqlite/Migrations/SqliteMigrationAnnotationProvider.cs [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+                try
+                {
+                    columnType = property.Sqlite().ColumnType ?? _typeMapper.GetDefaultMapping(property.ClrType).DefaultTypeName;
+                }
+                catch (NotSupportedException)
+                {
+                    columnType = null;
+                }
+                if (columnType != null
+                    && columnType.Equals("integer", StringComparison.CurrentCultureIgnoreCase)
```
Comment (bricelam) [inline src/EntityFramework.Sqlite/Migrations/SqliteMigrationAnnotationProvider.cs]: Also only if `property.StoreGeneratedPattern == StoreGeneratedPattern .Identity` This will let you turn it off in the case that the keys are managed by the application.

File: src/EntityFramework.Sqlite/Migrations/SqliteMigrationSqlGenerator.cs [hunk 10]
```diff
@@ -87,6 +89,68 @@ public virtual void Generate([NotNull] MoveDataOperation operation, [CanBeNull]
...
+                    .Append(" DEFAULT (")
+                    .Append(defaultExpression)
+                    .Append(")");
+            }
+            else if (defaultValue != null)
+            {
+                builder
+                    .Append(" DEFAULT ")
+                    .Append(_sql.GenerateLiteral((dynamic)defaultValue));
+            }
```
Comment (bricelam) [inline src/EntityFramework.Sqlite/Migrations/SqliteMigrationSqlGenerator.cs]: :frowning: Is there a way to not have to re-implement the base method? I think SQL Server calls base then tacks on `IDENTITY` to the end of the string.

File: test/EntityFramework.Sqlite.Tests/Migrations/SqliteMigrationSqlGeneratorTest.cs [hunk 11]
```diff
@@ -26,25 +27,86 @@ public void Insert_into_select()
             Assert.Equal("INSERT INTO \"RebuiltTable\" (\"col1\", \"col2\")" + EOL + "SELECT \"col1\", \"col2\" FROM \"OldTable\";" + EOL, Sql);
         }
 
-        [Fact]
-        public void CreateSchemaOperation_not_supported()
+        public override void CreateTableOperation()
         {
-            var ex = Assert.Throws<NotSupportedException>(() => Generate(new CreateSchemaOperation()));
-            Assert.Equal(Strings.SchemasNotSupported, ex.Message);
+            // replaced with test below
```
Comment (bricelam) [inline test/EntityFramework.Sqlite.Tests/Migrations/SqliteMigrationSqlGeneratorTest.cs]: Keep the original and add a new method instead? "Maintainable code" takes on a different meaning in test code--it's often OK to copy-and-paste in tests since it's better to be explicit about what's being tested in each case.

**Comment by bricelam** [comment]
:shipit: (once feedback is addressed)

File: src/EntityFramework.Sqlite/Migrations/SqliteMigrationAnnotationProvider.cs (hunk 8 above)
Comment (bricelam) [inline src/EntityFramework.Sqlite/Migrations/SqliteMigrationAnnotationProvider.cs]: Also, an auto-generated bool key sounds like fun: 1, 2, 3... becomes true, true, true... on the client

Comment (natemcmaster) [inline src/EntityFramework.Sqlite/Migrations/SqliteMigrationAnnotationProvider.cs]: I think technically it would work, but users may not expect this behavior if the CLR type isn't a integer-ish type (including short, long, and unsigned variants.)

File: src/EntityFramework.Sqlite/Migrations/SqliteMigrationSqlGenerator.cs (hunk 10 above)
Comment (natemcmaster) [inline src/EntityFramework.Sqlite/Migrations/SqliteMigrationSqlGenerator.cs]: SQLites is very specific about the order of words. 'integer [not null] autoincrement primary key' != 'integer primary key autoincrement not null'.

File: src/EntityFramework.Sqlite/Migrations/SqliteMigrationAnnotationProvider.cs (hunk 8 above)
Comment (natemcmaster) [inline src/EntityFramework.Sqlite/Migrations/SqliteMigrationAnnotationProvider.cs]: Summing up what we discussed: we will let users shoot themselves in the foot with autoincremented boolean. We only check now for StoreGeneratedPattern == Identity
